File: libs/dotnet/VietRide.Shared.Kernel/Primitives/QueryOptions.cs
namespace VietRide.Shared.Kernel.Primitives;

/// <summary>
/// Standard query-string parameters for list/collection endpoints (ADR 0004 §Decision Rule 6).
/// Bound from <c>?page=1&amp;pageSize=20&amp;search=...&amp;sortBy=createdAt&amp;sortDir=desc&amp;includeDeleted=false</c>.
/// </summary>
public sealed class QueryOptions
{
    private const string SortAscending = "asc";
    private const string SortDescending = "desc";

    private int _page = 1;
    private int _pageSize = 20;
    private string _sortDir = SortDescending;

    /// <summary>1-based page number. Defaults to 1.</summary>
    public int Page
    {
        get => _page;
        init => _page = value < 1 ? 1 : value;
    }

    /// <summary>Items per page. Clamped to 1–100 (BSOT §5.7). Defaults to 20.</summary>
    public int PageSize
    {
        get => _pageSize;
        init => _pageSize = value < 1 ? 1 : value > 100 ? 100 : value;
    }

    /// <summary>Free-text search term. Null/empty = no search.</summary>
    public string? Search { get; init; }

    /// <summary>
    /// Comma-separated list of fields to search in (e.g. <c>email,phone</c>).
    /// Whitelisted per aggregate at the repository layer (security requirement).
    /// </summary>
    public string? SearchIn { get; init; }

    /// <summary>Field to sort by. Whitelisted per aggregate at the repository layer.</summary>
    public string? SortBy { get; init; }

    /// <summary>Sort direction — <c>asc</c> or <c>desc</c>. Defaults to <c>desc</c>.</summary>
    public string SortDir
    {
        get => _sortDir;
        init => _sortDir = NormalizeSortDir(value);
    }

    /// <summary>
    /// When <c>true</c>, include soft-deleted records (ADR 0003).
    /// Only admin/privileged endpoints should expose this. Defaults to <c>false</c>.
    /// </summary>
    public bool IncludeDeleted { get; init; } = false;

    private static string NormalizeSortDir(string? value)
    {
        var normalized = value?.Trim().ToLowerInvariant();

        return normalized switch
        {
            SortAscending => SortAscending,
            SortDescending => SortDescending,
            _ => throw new ArgumentException("SortDir must be 'asc' or 'desc'.", nameof(value)),
        };
    }
}

File: libs/shared/contracts/src/dtos/query-options.ts
import { z } from 'zod';

/**
 * Shared query-string input for list/collection endpoints.
 * Per ADR 0004 Rule 6 / BSOT section 5.7.
 *
 * `sortBy` and `searchIn` are whitelist-validated per aggregate (security
 * requirement) — the repository/handler rejects any field not in its allow-list.
 */
export interface QueryOptions {
  /** 1-based page index. Default 1. */
  page: number;
  /** Items per page. Default 20, clamped 1..100. */
  pageSize: number;
  /** Free-text search term. */
  search?: string;
  /** Comma-separated fields to search in (whitelist per aggregate). */
  searchIn?: string;
  /** Field to sort by (whitelist per aggregate). */
  sortBy?: string;
  /** Sort direction. Default 'desc'. */
  sortDir: 'asc' | 'desc';
  /** Include soft-deleted records. Default false. */
  includeDeleted: boolean;
}

export const QueryOptionsSchema = z.object({
  page: z.coerce.number().int().min(1).default(1),
  pageSize: z.coerce
    .number()
    .int()
    .default(20)
    .transform((value) => Math.min(100, Math.max(1, value))),
  search: z.string().optional(),
  searchIn: z.string().optional(),
  sortBy: z.string().optional(),
  sortDir: z.enum(['asc', 'desc']).default('desc'),
  includeDeleted: z.coerce.boolean().default(false),
});
